### PR TITLE
gcp - add kubernetes initial setup

### PR DESCRIPTION
**Pull request by mediapills**
Add query and get methods for GKE node-pools and clusters.

### CODE REVIEW

File: tools/c7n_gcp/tests/data/events/k8s_create_cluster.json [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+            "failure-domain.beta.kubernetes.io/region": "us-central1",
+            "failure-domain.beta.kubernetes.io/zone": "us-central1-a",
+            "kubernetes.io/hostname": "gke-standard-cluster-1-pool-1-c733eefe-xtbq"
+          },
+          "name": "gke-standard-cluster-1-pool-1-c733eefe-xtbq"
+        },
+        "spec": {
+          "providerID": "gce://mythic-tribute-232915/us-central1-a/gke-standard-cluster-1-pool-1-c733eefe-xtbq"
+        },
+        "status": {
```
Comment (kapilt) [inline tools/c7n_gcp/tests/data/events/k8s_create_cluster.json]: please remove the status key in request and response, its extraneous and extremely large.

File: tools/c7n_gcp/tests/data/events/k8s_create_pool.json [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+          "labels": {
+            "component": "kube-proxy",
+            "tier": "node"
+          },
+          "name": "kube-proxy-gke-standard-cluster-1-pool-1-c733eefe-xtbq",
+          "namespace": "kube-system",
+          "selfLink": "/api/v1/namespaces/kube-system/pods/kube-proxy-gke-standard-cluster-1-pool-1-c733eefe-xtbq",
+          "uid": "9ef8eaa0a08f35eed9a7f7676a774e8a"
+        },
+        "spec": {
```
Comment (kapilt) [inline tools/c7n_gcp/tests/data/events/k8s_create_pool.json]: let's remove the spec key in request and response, there's hundreds of lines data here that we don't need.

File: tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    "ipAllocationPolicy": {},
+    "zone": "us-central1-a",
+    "servicesIpv4Cidr": "10.11.240.0/20",
+    "status": "RUNNING",
+    "currentNodeVersion": "1.11.8-gke.6",
+    "labelFingerprint": "a9dc16a7",
+    "currentMasterVersion": "1.11.8-gke.6",
+    "masterAuth": {
+      "username": "admin",
+      "password": "IWqLAzJg7bTJYP5M",
```
Comment (kapilt) [inline tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: this is kinda of shocking, and a security issue for gcp

File: tools/c7n_gcp/tests/test_gke.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        project_id = "cloud-custodian"
+
+        factory = self.replay_flight_data('gke-cluster-query', project_id)
+        p = self.load_policy(
+            {'name': 'all-gke-cluster',
+             'resource': 'gcp.gke-cluster'},
+            session_factory=factory
+        )
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
```
Comment (kapilt) [inline tools/c7n_gcp/tests/test_gke.py]: so we query massive amounts of data, but could return a {} and this test would still pass. please validate attributes on resources.

File: tools/c7n_gcp/tests/test_gke.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        project_id = "cloud-custodian"
+
+        factory = self.replay_flight_data('gke-cluster-nodepool-query', project_id)
+
+        p = self.load_policy(
+            {'name': 'all-gke-cluster-nodepools',
+             'resource': 'gcp.gke-cluster-nodepool'},
+            session_factory=factory)
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
```
Comment (kapilt) [inline tools/c7n_gcp/tests/test_gke.py]: validate a resource attribute here as well.

File: tools/c7n_gcp/c7n_gcp/resources/gke.py [hunk 6]
```diff
@@ -11,19 +11,77 @@
...
+                'get', verb_arguments={
+                    'name': 'projects/{}/locations/{}/clusters/{}'.format(
+                        resource_info['project_id'],
+                        resource_info['location'],
+                        resource_info['cluster_name'])})
+
+
+@resources.register('gke-cluster-nodepool')
+class KubernetesClusterNodePool(ChildResourceManager):
+
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/gke.py]: please add class doc strings with the url to the gcp rest api docs for the resource

File: tools/c7n_gcp/c7n_gcp/resources/gke.py [hunk 7]
```diff
@@ -11,19 +11,77 @@
...
         service = 'container'
-        version = 'v1beta1'
+        version = 'v1'
         component = 'projects.locations.clusters'
         enum_spec = ('list', 'clusters[]', None)
         scope = 'project'
         scope_key = 'parent'
         scope_template = "projects/{}/locations/-"
         id = "name"
+
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/gke.py]: please add class doc strings with the url to the gcp rest api docs for the resource

File: tools/c7n_gcp/c7n_gcp/resources/gke.py [hunk 8]
```diff
@@ -11,19 +11,77 @@
...
+        def get(client, resource_info):
+            return client.execute_query(
+                'get', verb_arguments={
+                    'name': 'projects/{}/locations/{}/clusters/{}'.format(
+                        resource_info['project_id'],
+                        resource_info['location'],
+                        resource_info['cluster_name'])})
+
+
+@resources.register('gke-cluster-nodepool')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/gke.py]: rename to just gke-nodepool   cluster is implied, ie there isn't any other kind.

File: tools/c7n_gcp/tests/data/events/k8s_create_cluster.json (hunk 1 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/data/events/k8s_create_cluster.json]: Updated

File: tools/c7n_gcp/tests/data/events/k8s_create_pool.json (hunk 2 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/data/events/k8s_create_pool.json]: Updated

File: tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json (hunk 3 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: Updated

File: tools/c7n_gcp/tests/test_gke.py (hunk 4 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/test_gke.py]: Updated

File: tools/c7n_gcp/tests/test_gke.py (hunk 5 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/test_gke.py]: Updated

File: tools/c7n_gcp/c7n_gcp/resources/gke.py (hunk 6 above)
Comment (Dkhanas) [reply tools/c7n_gcp/c7n_gcp/resources/gke.py]: Updated

File: tools/c7n_gcp/c7n_gcp/resources/gke.py (hunk 7 above)
Comment (Dkhanas) [reply tools/c7n_gcp/c7n_gcp/resources/gke.py]: Updated

File: tools/c7n_gcp/c7n_gcp/resources/gke.py (hunk 8 above)
Comment (Dkhanas) [reply tools/c7n_gcp/c7n_gcp/resources/gke.py]: Updated

File: tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json (hunk 3 above)
Comment (kapilt) [reply tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: the fact that master passwords are being logged is something that needs to be flagged to gcp folks at google as a security issue.

Comment (kapilt) [reply tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: oh this isn't even logging isn't on api trail, this is part of a get request for a cluster.. hmmm okay. so read access to a project means root access to your cluster.. wtf.

**Comment by kapilt** [comment]
There's thousands of line of json here, that don't feel material to the functionality being tested which is bloating this pr. Please try to slim them down, from initial comments about that to last push, the pr got bigger with even more extraneous json.

File: tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json (hunk 3 above)
Comment (Dkhanas) [reply tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: This section has been deleted in a mock.

Comment (kapilt) [reply tools/c7n_gcp/tests/data/flights/gke-cluster-nodepool-query/get-v1-projects-cloud-custodian-locations-us-central1-a-clusters-standard-cluster-1-nodePools_1.json]: sorry the wtf wasn't in ref to the pr, more to just that this passwords were being returned from read apis at all from the gcp api. i went ahead and talked to some gcp folks, and was told this was a legacy auth mechanism for older clusters that is now deprecated.

**Pull request merged**